lib/active_job/queue_adapters/pubsub_adapter.rb

```diff
@@ -4,18 +4,24 @@ module ActiveJob
   module QueueAdapters
     class PubsubAdapter
       # Enqueue a job to be performed.
-      #
       # @param [ActiveJob::Base] job The job to be performed.
       def enqueue(job)
-        raise(NotImplementedError)
+        Rails.logger.info("[PubsubAdapter] Enqueue: #{job.arguments[0]}")
+        Base.execute(job.serialize)
       end
 
       # Enqueue a job to be performed at a certain time.
-      #
       # @param [ActiveJob::Base] job The job to be performed.
       # @param [Float] timestamp The time to perform the job.
       def enqueue_at(job, timestamp)
-        raise(NotImplementedError)
+        Rails.logger.info("[PubsubAdapter] Enqueue At: #{job.arguments[0]} - #{Time.at(timestamp)}")
+        delay = timestamp - Time.current.to_f
+
+        if delay.positive?
+          Concurrent::ScheduledTask.execute(delay) { Base.execute(job.serialize) }
+        else
+          Base.execute(job.serialize)
+        end
```

Comment:
This will lead to loss of jobs during server restarts and crashes.

Reply:
True. I knew this part of the assessment is the area that needs most improving but failed to do so.

Again, I know the assessment is done but for learning purposes, I would appreciate if you can suggest an approach of how I can preserve scheduled jobs on server restarts and failures.

Reply:
One way might be to push jobs to Pub/Sub and keep postponing the acknowledgment deadline until messages are ready for execution.

Reply:
I see. So an `execute_at` attribute:

```
if current_time < data['execute_at']
  // push to Pub/Sub again via job
  return
end

received_message.acknowledge!
```

Thanks for answering. 💯